Translate React (top-level) API Reference (#47)
reactjs/ja.reactjs.org: content/docs/reference-react.md

  - "docs/top-level-api-zh-CN.html"
---

`React` は React ライブラリのエントリポイントです。`<script>` タグから React を読み込む場合、これらの最上位 API をグローバルの `React` から利用できます。npm と ES6 を使う場合、`import React from 'react'` と書けます。npm と ES5 を使う場合、`var React = require('react')` と書けます。

## Overview

### Components

React コンポーネントを使用すると UI を独立した再利用可能な部分に分割し、各部分を個別に考えることができます。React コンポーネントは `React.Component` または `React.PureComponent` をサブクラス化することで定義できます。

 - [`React.Component`](#reactcomponent)
 - [`React.PureComponent`](#reactpurecomponent)

ES6 クラスを使わない場合は、代わりに `create-react-class` モジュールを使うことができます。詳しくは [Using React without ES6](/docs/react-without-es6.html) を参照してください。

React コンポーネントは関数で定義でき、その際に以下の関数でラップできます：

- [`React.memo`](#reactmemo)

### Creating React Elements

UI がどのように見えるべきかを記述するために [JSX の使用](/docs/introducing-jsx.html) を推奨します。JSX のそれぞれの要素は [`React.createElement()`](#createelement) を呼ぶための単なる糖衣構文です。JSX を使用している場合は、通常、次のメソッドを直接呼び出すことはありません。

- [`createElement()`](#createelement)
- [`createFactory()`](#createfactory)

> 詳しくは [Using React without JSX](/docs/react-without-jsx.html) を参照してください。

Reviewer: ```suggestion
詳しくは [JSX なしで React を使う](/docs/react-without-jsx.html) を参照してください。
```

  - "docs/top-level-api-zh-CN.html"
---

`React` は React ライブラリのエントリポイントです。`<script>` タグから React を読み込む場合、これらの最上位 API をグローバルの `React` から利用できます。
npm と ES6 を使う場合、 `import React from 'react'` と書けます。npm と ES5 を使う場合、 `var React = require('react')` と書けます。

## Overview

### Components

> React コンポーネントを使用するとUIを独立した再利用可能な部分に分割し、各部分を個別に考えることができます。

Reviewer 2: UI の部分ですが、半角スペースを空けるようにお願いします！
他についても、同様です。

React コンポーネントは `React.Component` または `React.PureComponent` をサブクラス化することで定義することができます。

 - [`React.Component`](#reactcomponent)
 - [`React.PureComponent`](#reactpurecomponent)

ES6 クラスを使う代わりに、 `create-react-class` を使うことができます。

> 詳しくは [Using React without ES6](/docs/react-without-es6.html) を見てください。

Reviewer 2: [imo]
参照くださいでも良いかもしれませんね。

Author: "見てください" から "参照してください" に変更しました。

- [`React.createRef`](#reactcreateref)
- [`React.forwardRef`](#reactforwardref)

### サスペンス (Suspense)

サスペンスを使用すると、コンポーネントはレンダーの前に何かを「待機」できます。現在、サスペンスは 1 つのユースケースのみをサポートしています：[`React.lazy` を使ってコンポーネントを動的に読み込む](/docs/code-splitting.html#reactlazy)。将来的にはデータの取得のような他のユースケースもサポートされるでしょう。

- [`React.lazy`](#reactlazy)
- [`React.Suspense`](#reactsuspense)

* * *

> ## レファレンス

Reviewer: ```suggestion
## リファレンス
```

- [`React.createRef`](#reactcreateref)
- [`React.forwardRef`](#reactforwardref)

### サスペンス (Suspense)

サスペンスを使用すると、コンポーネントはレンダーの前に何かを「待機」することができます。
現在、サスペンスは 1 つのユースケースのみをサポートしています： [loading components dynamically with `React.lazy`](/docs/code-splitting.html#reactlazy)。

> 将来的にはデータの取得のような他のユースケースもサポートされます。

Reviewer 3: ```suggestion
将来的にはデータの取得のような他のユースケースもサポートされるでしょう。
```

- [`React.createRef`](#reactcreateref)
- [`React.forwardRef`](#reactforwardref)

> ### サスペンス

Reviewer 2: [note]
用語集に追加の必要あり？

> サスペンスを使用すると、コンポーネントはレンダリングの前に何かを待機することができます。

Reviewer 2: 初見の概念だと思うので、サスペンス(Suspense)とかけると良いかと！

}
```

基底クラス `React.Component` に関するメソッドとプロパティの一覧については、[React.Component API Reference](/docs/react-component.html) を参照してください。

* * *

### `React.PureComponent`

> `React.PureComponent` は [`React.Component`](#reactcomponent) と似ています。両者の違いは [`React.Component`](#reactcomponent) が [`shouldComponentUpdate()`](/docs/react-component.html#shouldcomponentupdate) を実装していないことに対し、`React.PureComponent` は props と state を浅く (shallow) 比較することでそれを実装していることです。

Reviewer: ```suggestion
`React.PureComponent` は [`React.Component`](#reactcomponent) と似ています。両者の違いは [`React.Component`](#reactcomponent) が [`shouldComponentUpdate()`](/docs/react-component.html#shouldcomponentupdate) を実装していないのに対し、`React.PureComponent` は props と state を浅く (shallow) 比較することでそれを実装していることです。
```

}
```

基底クラス `React.Component` に関するメソッドとプロパティの一覧については、[React.Component API Reference](/docs/react-component.html) を参照してください。

* * *

### `React.PureComponent`

`React.PureComponent` は [`React.Component`](#reactcomponent) と似ています。両者の違いは [`React.Component`](#reactcomponent) が [`shouldComponentUpdate()`](/docs/react-component.html#shouldcomponentupdate) を実装していないことに対し、`React.PureComponent` は props と state を浅く (shallow) 比較することでそれを実装していることです。

> React コンポーネントの `render()` 関数が同じ props と state を与えられたときに同じ結果をレンダリングするときは、パフォーマンスを向上させるために `React.PureComponent` を使用することができます。

Reviewer 3: ```suggestion
もしReact コンポーネントの `render()` 関数が同じ props と state を与えられたときに同じ結果をレンダリングするなら、パフォーマンスを向上させるために `React.PureComponent` を使用することができます。
```

}
```

基底クラス `React.Component` に関するメソッドとプロパティの一覧については、[React.Component API Reference](/docs/react-component.html) を参照してください。

* * *

### `React.PureComponent` {#reactpurecomponent}

`React.PureComponent` は [`React.Component`](#reactcomponent) と似ています。両者の違いは [`React.Component`](#reactcomponent) が [`shouldComponentUpdate()`](/docs/react-component.html#shouldcomponentupdate) を実装していないことに対し、`React.PureComponent` は props と state を浅く (shallow) 比較することでそれを実装していることです。

React コンポーネントの `render()` 関数が同じ props と state を与えられたときに同じ結果をレンダーするときは、パフォーマンスを向上させるために `React.PureComponent` を使用できます。

> 補足
>

> > `React.PureComponent` の `shouldComponentUpdate()` は オブジェクトの浅い比較のみを行います。これらに複雑なデータ構造が含まれていると、深い部分のみに差分があるために、本当は差分があるにも関わらず差分がないと見なされる場合があります。単純な props とstate を持つ場合にのみ `PureComponent` を継承するか、深いデータ構造が変更されたとわかっているときに [`forceUpdate()`](/docs/react-component.html#forceupdate) を使用してください。あるいは、ネストされたデータ構造の高速な比較を容易にするために [イミュータブルなオブジェクト](https://facebook.github.io/immutable-js/) の使用を検討してください。

Reviewer: ```suggestion
> `React.PureComponent` の `shouldComponentUpdate()` は オブジェクトの浅い比較のみを行います。これらに複雑なデータ構造が含まれていると、深い部分のみに差分があるために、本当は差分があるにも関わらず差分がないと見なされる場合があります。単純な props と state を持つ場合にのみ `PureComponent` を継承するか、深いデータ構造が変更されたとわかっているときに [`forceUpdate()`](/docs/react-component.html#forceupdate) を使用してください。あるいは、ネストされたデータ構造の高速な比較を容易にするために [イミュータブルなオブジェクト](https://facebook.github.io/immutable-js/) の使用を検討してください。
```

});
```

> `React.memo` は [高階コンポーネント](/docs/higher-order-components.html) です。これは [`React.PureComponent`](#reactpurecomponent) に似ていますが、クラスではなく関数コンポーネントです。

Reviewer: ```suggestion
`React.memo` は [高階コンポーネント](/docs/higher-order-components.html) です。これは [`React.PureComponent`](#reactpurecomponent) に似ていますが、クラスではなく関数コンポーネント用です。
```

"for" の訳抜けがあります。`React.memo` 自体はもちろんクラスではありませんが、関数コンポーネントでもありません。`React.memo` 自体は、「関数コンポーネントを受け取って関数を返す関数」です。

});
```

`React.memo` は [高階コンポーネント](/docs/higher-order-components.html) です。これは [`React.PureComponent`](#reactpurecomponent) に似ていますが、クラスではなく関数コンポーネントです。

> ある関数コンポーネントが同じ props を与えられたときに同じ結果をレンダーするとき、結果を記憶してパフォーマンスを向上させるためにそれを `React.memo` でラップして呼び出すことができます。つまり、Reactはコンポーネントのレンダーをスキップし、最後のレンダー結果を再利用します。

Reviewer 3: ```suggestion
もしある関数コンポーネントが同じ props を与えられたときに同じ結果をレンダーするなら、結果を記憶してパフォーマンスを向上させるためにそれを `React.memo` の呼び出しでラップすることができます。つまり、Reactはコンポーネントのレンダーをスキップし、最後のレンダー結果を再利用します。
```

Reviewer: ```suggestion
もしある関数コンポーネントが同じ props を与えられたときに同じ結果をレンダーするなら、結果を記憶してパフォーマンスを向上させるためにそれを `React.memo` でラップすることができます。つまり、Reactはコンポーネントのレンダーをスキップし、最後のレンダー結果を再利用します。
```

「呼び出し」の正しい訳し方を考えるよりシンプルに済ませてよさそう :)

Reviewer 3: なぜシンプルに済ませて良いのですか？
「ラップする」と「ラップして呼び出す」では意味が違うと思いますが。

Reviewer: 1. 原文： wrap it in a call to `React.memo`
2. 超直訳：それを `React.memo` の呼び出しの中でラップする
3. 私の元の翻訳：それを `React.memo` でラップして呼び出す
4. takanoripさん提案：それを `React.memo` の呼び出しでラップする
5. 別にこれでもいいのでは：それを `React.memo` でラップする

3 は直訳ではない（ラップされた関数を呼び出す話は確かに原文に一切ない）けど、説明としては読みやすい。4 は直訳に近いですが個人的に見慣れない表現で、これだったらいっそ 5 でも同じで、かつより素直な表現なのでは、という気がしました。別案としては「`React.memo` を呼び出してラップする」もありだと思います。

デフォルトでは props オブジェクト内の複雑なオブジェクトは浅い比較のみが行われます。比較を制御したい場合は 2 番目の引数でカスタム比較関数を指定することができます。

```javascript
function MyComponent(props) {
  /* render using props */
}
function areEqual(prevProps, nextProps) {
  /*
  nextPropsをrenderに渡したときに

>   prevPropsをrenderに渡したときと同じ結果になるときにtrue

Reviewer 3: ```suggestion
  prevPropsをrenderに渡したときと同じ結果になるときにtrueを返す
```

)
```

> `element` から新しい React 要素を複製して返します。結果の要素は元の要素の props と新しい props が浅くマージされたものを持ちます。新しい小要素は既存の小要素を置き換えます。`key` と `ref` は元の要素から保持されます。

Reviewer 3: ```suggestion
`element` から新しい React 要素を複製して返します。結果の要素は元の要素の props と新しい props が浅くマージされたものを持ちます。新しい子要素は既存の子要素を置き換えます。`key` と `ref` は元の要素から保持されます。
```

`React.cloneElement()` は以下のコードとほぼ同等です：

```js
<element.type {...element.props} {...props}>{children}</element.type>
```

> ただし、`ref` も保持されます。つまり `ref` のある子要素を得ても、間違って祖先から誤ってそれを盗むことはありません。新しい要素にも同じ `ref` が取り付けられます。

Reviewer 3: ```suggestion
ただし、`ref` も保持されます。つまり `ref` のある子要素を受け取っても、間違って元の React 要素から  `ref` を盗むことはありません。新しい要素にも同じ `ref` が追加されます。
```

React.createFactory(type)
```

与えられた型の React 要素を生成する関数を返します。[`React.createElement()`](#createElement) と同様に、`type` 引数はタグ名の文字列（`'div'` や `'span'` など）、[React コンポーネント](/docs/components-and-props.html)型（クラスもしくは関数）、[React フラグメント](#reactfragment)型のいずれかです。

> このヘルパーはレガシーと見なされます。JSX か `React.createElement()` を直接使用することを勧めます。

Reviewer 3: ```suggestion
このヘルパーはレガシーだと考えられているので、代わりにJSX か `React.createElement()` を直接使用することをおすすめします。
```

React.createFactory(type)
```

> 与えられた型の React 要素を生成する関数を返します。[`React.createElement()`](#createElement) と同様に、type 引数はタグ名の文字列（`div` や `span` など）、[React component](/docs/components-and-props.html) 型（クラスもしくは関数）、[React fragment](#reactfragment) 型のいずれかです。

Reviewer: ```suggestion
与えられた型の React 要素を生成する関数を返します。[`React.createElement()`](#createElement) と同様に、`type` 引数はタグ名の文字列（`'div'` や `'span'` など）、[React コンポーネント](/docs/components-and-props.html)型（クラスもしくは関数）、[React フラグメント](#reactfragment)型のいずれかです。
```

このヘルパーはレガシーと見なされます。JSX か `React.createElement()` を直接使用することを勧めます。

> JSX を使っていれば通常  `React.createFactory()` を直接呼び出すことはありません。詳しくは [React Without JSX](/docs/react-without-jsx.html) を見てください。

Reviewer: ```suggestion
JSX を使っていれば通常 `React.createFactory()` を直接呼び出すことはありません。詳しくは [JSX なしで React を使う](/docs/react-without-jsx.html) を参照してください。
```

React.isValidElement(object)
```

オブジェクトが React 要素であることを確認します。`true` または `false` を返します。

* * *

### `React.Children`

> `React.Children` はデータ構造が非公開の `this.props.children` を扱うためのユーティリティを提供します。

Reviewer 3: ```suggestion
`React.Children` はデータ構造が不明瞭な `this.props.children` を扱うためのユーティリティを提供します。
```

Reviewer: 「データ構造が不明瞭な」だと訳が分からないので、これは「非公開」でいいと思いますが。もちろん React の中の人はデータ構造を知っているわけですし。 `console.log(props.children)` とかすれば中身は明瞭に見えます（ドキュメント化されていないReactの仮想DOMオブジェクト本体です）。

#### `React.Children.map`

```javascript
React.Children.map(children, function[(thisArg)])
```

`this` を `thisArg` に設定して、`children` 内に含まれるすべての直下の子要素に対して関数を呼び出します。`children` が配列の場合は走査され、配列の各要素に対して関数が呼び出されます。`children` が `null` または `undefined` の場合はこのメソッドは配列ではなく `null` または `undefined` を返します。

> 補足
>
> `children` が `Fragment` の場合、それは 1 つの子要素として扱われ、走査されません。

#### `React.Children.forEach`

```javascript
React.Children.forEach(children, function[(thisArg)])
```

[`React.Children.map()`](#reactchildrenmap) と似ていますが、配列を返しません。

#### `React.Children.count`

```javascript
React.Children.count(children)
```

> `children` のコンポーネントの数を返します。これは `map` または `forEach` に渡したコールバックが呼ばれる回数と同じです。

Reviewer 3: ```suggestion
`children` に含まれるコンポーネントの総数を返します。これは `map` または `forEach` に渡したコールバックが呼ばれる回数と同じです。
```

#### `React.Children.only`

```javascript
React.Children.only(children)
```

> `children` が 1 つの子要素しか持たないことを確認します。そうでない場合、このメソッドはエラーを投げます。

Reviewer 3: ```suggestion
`children` が 1 つの子要素しか持たないことを確認しし、その結果を返します。そうでない場合、このメソッドはエラーを投げます。
```

React.isValidElement(object)
```

> オブジェクトが React 要素であることを確認します。 `true` または `false` を返します。

Reviewer: ```suggestion
オブジェクトが React 要素であることを確認します。`true` または `false` を返します。
```

* * *

### `React.Children`

> `React.Children` はデータ構造が不透明な `this.props.children` を扱うためのユーティリティを提供します。

Reviewer: ```suggestion
`React.Children` はデータ構造が非公開の `this.props.children` を扱うためのユーティリティを提供します。
```

opaque を更に意訳しました。実際 children は `{$$typeof: Symbol(react.element), type: "div", key: ".$apple", ref: null, props: {…}, …}` みたいな謎の形をしているオブジェクトの配列です。

#### `React.Children.map`

```javascript
React.Children.map(children, function[(thisArg)])
```

> `this` を `thisArg` に設定して、 `children` 内に含まれるすべての直下の子要素に対して関数を呼び出します。 `children` が配列の場合は走査され、配列の各要素に対して関数が呼び出されます。 `children` が `null` または `undefined` の場合はこのメソッドは配列ではなく `null` または `undefined` を返します。

Reviewer: ```suggestion
`this` を `thisArg` に設定して、`children` 内に含まれるすべての直下の子要素に対して関数を呼び出します。`children` が配列の場合は走査され、配列の各要素に対して関数が呼び出されます。`children` が `null` または `undefined` の場合はこのメソッドは配列ではなく `null` または `undefined` を返します。
```

> 補足
>

> > `children` が `Fragment` の場合、それは1つの子要素として扱われ、走査されません。

Reviewer: ```suggestion
> `children` が `Fragment` の場合、それは 1 つの子要素として扱われ、走査されません。
```

#### `React.Children.forEach`

```javascript
React.Children.forEach(children, function[(thisArg)])
```

[`React.Children.map()`](#reactchildrenmap) と似ていますが、配列を返しません。

#### `React.Children.count`

```javascript
React.Children.count(children)
```

`children` のコンポーネントの数を返します。これは `map` または `forEach` に渡したコールバックが呼ばれる回数と同じです。

#### `React.Children.only`

```javascript
React.Children.only(children)
```

> `children` が1つの子要素しか持たないことを確認します。そうでない場合、このメソッドはエラーを投げます。

Reviewer: ```suggestion
`children` が 1 つの子要素しか持たないことを確認します。そうでない場合、このメソッドはエラーをスローします。
```

> 補足:
>
> [`React.Children.map()`](#reactchildrenmap) の返り値は React 要素ではなく配列なため、 `React.Children.only()` はそれを受け付けません。

#### `React.Children.toArray`

```javascript
React.Children.toArray(children)
```

> それぞれの要素に割り当てられたキーを持つ不透明なデータ構造の `children` を平坦な配列として返します。レンダリングメソッド内で子の集合を操作したい場合、特に `this.props.children` を渡す前に並べ替えたりスライスしたい場合に便利です。

Reviewer: ```suggestion
不透明なデータ構造である `children` を平坦な配列として返し、それぞれの要素に key を割り当てます。レンダーメソッド内で子の集合を操作したい場合、特に `this.props.children` を渡す前に並べ替えたりスライスしたい場合に便利です。
```

よく分からなかったので例を https://codepen.io/anon/pen/VgbgpL に作ってみました。return array の item には元の key に接頭辞がついたものが割り当てられるということなんですね…

> 補足:
>

> > `React.Children.toArray()` は子のリストを平坦にするときにネストされた配列の意味を保つためにキーを変更します。つまり、 `toArray` は配列のそれぞれの要素の key に接頭辞を付けて返します。

Reviewer: ```suggestion
> `React.Children.toArray()` は子のリストを平坦にするときにネストされた配列の意味を保つために key を変更します。つまり、`toArray` は配列のそれぞれの要素の key に接頭辞を付けて返します。
```

### `React.createRef`

`React.createRef` は [ref](/docs/refs-and-the-dom.html) を作成します。ref は ref 属性を介して React 要素に取り付けることができます。

### `React.forwardRef`

`React.forwardRef` は [ref](/docs/refs-and-the-dom.html) を配下のツリーの別のコンポーネントに受け渡す React コンポーネントを作成します。この手法はあまり一般的ではありませんが、2 つのシナリオで特に役立ちます：

* [Forwarding refs to DOM components](/docs/forwarding-refs.html#forwarding-refs-to-dom-components)
* [Forwarding refs in higher-order-components](/docs/forwarding-refs.html#forwarding-refs-in-higher-order-components)

`React.forwardRef` はレンダリング関数を引数として受け入れます。Reactは props と ref を 2 つの引数として呼び出します。この関数は React ノードを返す必要があります。

`embed:reference-react-forward-ref.js`

上の例では、React は `<FancyButton ref={ref}>` 要素に与えた `ref` を `React.forwardRef` の呼び出し内のレンダリング関数の 2 番目の引数として渡します。このレンダリング関数は `ref` を `<button ref={ref}>` 要素に渡します。

結果として、Reactが `ref` を取り付けた後、`ref.current` は `<button>` の DOM 要素のインスタンスを直接指すようになります。

詳しくは [forwarding refs](/docs/forwarding-refs.html) を参照してください。

### `React.lazy`

`React.lazy()` を使用すると、動的に読み込まれるコンポーネントを定義することができます。これにより、バンドルサイズを削減して、最初のレンダリング時に使用されないコンポーネントの読み込みを遅らせることができます。

> [code splitting のドキュメント](/docs/code-splitting.html#reactlazy)から使用方法を学ぶことが出来ます。また、使い方をより詳しく説明した[こちらの記事](https://medium.com/@pomber/lazy-loading-and-preloading-components-in-react-16-6-804de091c82d)もチェックしてみてください。

Reviewer 2: 「出来ます」は可能な限り漢字をひらく形で、「できます」とするほうが望ましいと思います。

このあたりは Lint 的なので prh でルール化したいですね

Author: ありがとうございます。癖で「出来ます」としがちなので「できます」に置換した覚えがあるのですが残っていたみたいです。修正します。

Author: これ「出来る」を「できる」にしただけだったようです・・・。

```js
// This component is loaded dynamically
const SomeComponent = React.lazy(() => import('./SomeComponent'));
```

`lazy` コンポーネントをレンダリングするには `<React.Suspense>` がレンダリングツリーの上位に必要です。これはローディングインジケータを指定する方法です。

> **補足**
>
> `React.lazy` を使って動的にインポートするには JS 環境で Promise が使用できる必要があります。これは IE11 以前の環境ではポリフィルが必要だということです。

### `React.Suspense`

> `React.Suspense` を使用することで、その配下にレンダリングする準備ができていないコンポーネントがあるときにローディングインジケータを指定することができます。現在、遅延読み込みコンポーネントは `<React.Suspense>` のみによってサポートされています。

Reviewer 2: 「できます」系ですが、「することができます」は冗長であり、意味としては「できます」から情報量が増えていないことから、「することが」を除くことが一般的に推奨されています。

どうでしょうか？

Author: ありがとうございます。こちらも同様に修正します。

}
```

> また、フラグメントを `<></>` という短縮構文で使用することもできます。詳しくは [React v16.2.0: Improved Support for Fragments](/blog/2017/11/28/react-v16.2.0-fragment-support.html) を見てください。

Reviewer: ```suggestion
また、フラグメントを `<></>` という短縮構文で使用することもできます。詳しくは [React v16.2.0: Improved Support for Fragments](/blog/2017/11/28/react-v16.2.0-fragment-support.html) を参照してください。
```

### `React.createRef`

`React.createRef` は [ref](/docs/refs-and-the-dom.html) を作成します。ref は ref 属性を介して React 要素に取り付けることができます。

### `React.forwardRef`

> `React.forwardRef` は [ref](/docs/refs-and-the-dom.html) を配下のツリーの別のコンポーネントに受け渡す React コンポーネントを作成します。この手法はあまり一般的ではありませんが、2つのシナリオで特に役立ちます：

Reviewer: ```suggestion
`React.forwardRef` は [ref](/docs/refs-and-the-dom.html) を配下のツリーの別のコンポーネントに受け渡す React コンポーネントを作成します。この手法はあまり一般的ではありませんが、2 つのシナリオで特に役立ちます：
```

* [Forwarding refs to DOM components](/docs/forwarding-refs.html#forwarding-refs-to-dom-components)
* [Forwarding refs in higher-order-components](/docs/forwarding-refs.html#forwarding-refs-in-higher-order-components)

`React.forwardRef` はレンダリング関数を引数として受け入れます。Reactは props と ref を2つの引数として呼び出します。この関数は React ノードを返す必要があります。

`embed:reference-react-forward-ref.js`

上の例では、React は `<FancyButton ref={ref}>` 要素に与えた `ref` を `React.forwardRef` の呼び出し内のレンダリング関数の2番目の引数として渡します。このレンダリング関数は `ref` を `<button ref={ref}>` 要素に渡します。

結果として、Reactが `ref` を取り付けた後、`ref.current` は `<button>` の DOM 要素のインスタンスを直接指すようになります。

詳しくは [forwarding refs](/docs/forwarding-refs.html) を見てください。

### `React.lazy`

> `React.lazy()` を使用すると、動的に読み込まれるコンポーネントを定義することができます。これにより、バンドルサイズを削減して、最初のレンダリング時に使用されなかったコンポーネントの読み込みを遅らせることができます。

Reviewer: ```suggestion
`React.lazy()` を使用すると、動的に読み込まれるコンポーネントを定義することができます。これにより、バンドルサイズを削減して、最初のレンダリング時に使用されないコンポーネントの読み込みを遅らせることができます。
```

> [code splitting documentation](/docs/code-splitting.html#reactlazy) から使用方法を学ぶことが出来ます。また、使い方をより詳しく説明した [this article](https://medium.com/@pomber/lazy-loading-and-preloading-components-in-react-16-6-804de091c82d) もチェックしてみてください。

Reviewer: ```suggestion
[code splitting のドキュメント](/docs/code-splitting.html#reactlazy)から使用方法を学ぶことが出来ます。また、使い方をより詳しく説明した[こちらの記事](https://medium.com/@pomber/lazy-loading-and-preloading-components-in-react-16-6-804de091c82d)もチェックしてみてください。
```

```js
// This component is loaded dynamically
const SomeComponent = React.lazy(() => import('./SomeComponent'));
```

`lazy` コンポーネントをレンダリングするには `<React.Suspense>` がレンダリングツリーの上位に必要です。これはローディングインジケータを指定する方法です。

> **補足**
>

> > `React.lazy` を使って動的にインポートするには Promise が JS 環境で使用できる必要があります。これはIE11以前の環境にポリフィルを要求します。

Reviewer: ```suggestion
> `React.lazy` を使って動的にインポートするには JS 環境で Promise が使用できる必要があります。これは IE11 以前の環境ではポリフィルが必要だということです。
```

}
```

> これは [code splitting guide](/docs/code-splitting.html#reactlazy) で文書化されています。遅延コンポーネントを `Suspense` ツリーの奥深くに置くこともできますが、それらすべてをラップする必要はありません。ベストプラクティスは `<Suspense>` をローディングインジケータを表示したい場所に配置することですが、コードを分割したい場合は `lazy()` を使用してください。

Reviewer: ```suggestion
これは [code splitting の解説](/docs/code-splitting.html#reactlazy)で文書化されています。遅延される (lazy) コンポーネントを `Suspense` ツリーの奥深くに置くことができ、それらを 1 つずつラップする必要はありません。ベストプラクティスは `<Suspense>` をローディングインジケータを表示したい場所に配置して、コードを分割したいところで常に `lazy()` を使用する、ということです。
```

「ですが」を使うと逆に伝わりづらいと感じました

> これらは現在サポートされていませんが、将来的には `Suspense` にデータの取得などのより多くのシナリオを処理させる予定です。この計画について、 [our roadmap](/blog/2018/11/27/react-16-roadmap.html) を読むことができます。

Reviewer: ```suggestion
これらは現在サポートされていませんが、将来的には `Suspense` にデータの取得などのより多くのシナリオを処理させる予定です。これについては[ロードマップ](/blog/2018/11/27/react-16-roadmap.html)で読めます。
```
